service/lib/dinstaller/progress.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2022] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+module DInstaller
+  # Class to pass around to get progress info. Intention is to allow
+  # in long run methods like probe or commit to pass its progress report
+  # when UI asks for it.
+  #
+  # The class contain major and minor steps. Intention is that
+  # top level manager knows about major steps and sets total and current step.
+  # On other hand each step called by manager class sets minor steps and messages if needed.
+  #
+  # @example manager interaction
+  #
+  #   progress = Progress.new
+  #   progress.init_progress(3, "Doing step1")
+  #   step1(progress)
+  #   progress.next_step("Doing step2")
+  #   step2(progress)
+  #   progress.next_step("Doing step3")
+  #   step3(progress)
+  #   progress.next_step("Finished with step3")
+  #
+  # @example module interaction
+  #
+  #   def step2(progress)
+  #     progress.init_minor_steps(2, "Doing subtask1")
+  #     subtask1
+  #     progress.next_minor_step("Doing subtask2")
+  #     subtask2
+  #     progress.next_minor_step("Finished subtask2")
+  class Progress
```

Comment:
Maybe it would too much, but I also imagine this as a kind of `ProgressManager`, generating steps when requested:

~~~
steps_manager = StepsManager.new(steps: 3)
steps_manager.next_step("Doing step1") #=> DInstaller::Step
do_step1(steps_manager)
steps_manager.next_step("Doing step2")
do_step2(steps_manager)
steps_manager.next_step("Doing step3")
do_step3(steps_manager)
...

def do_step1(steps_manager)
  current_step = steps_manager.current_step
  sub_steps_manager = current_step.sub_steps(2) #=> DInstaller::StepsManager
  sub_steps_manager.next_step("Doing" subtask1")
  subtask1
  sub_steps_manager.next_step("Doing" subtask2")
  subtask2
~~~ 

This would allow to generate a tree of steps: step1 -> step 1.1 -> step 1.1.1. With this, we could also have different callbacks for substeps, etc. But as said, maybe is too much.

Reply:
Well, having tree of steps is something that is hard to imagine for me how to display it in UI. I try to keep it simple now. So top level manager knows about top level steps and then module knows about module specific substeps if needed. ( from my current testing some steps like language probe is super fast, that it is even hard to notice it ).